src/core/thread/mesh_forwarder.cpp

```diff
@@ -1596,6 +1634,69 @@ otError MeshForwarder::GetFramePriority(uint8_t *           aFrame,
     return error;
 }
 
+#if OPENTHREAD_FTD
+Message *MeshForwarder::FindFragmentPriorityMessage(Lowpan::FragmentHeader &aFragmentHeader)
+{
+    Message *message;
+
+    for (message = mFragmentPriorityQueue.GetHead(); message; message = message->GetNext())
+    {
+        if (message->GetDatagramTag() == aFragmentHeader.GetDatagramTag())
+        {
+            break;
+        }
+    }
+
+    return message;
+}
+
+otError MeshForwarder::GetFragmentPriority(Lowpan::FragmentHeader &aFragmentHeader, uint8_t &aPriority)
+{
+    otError  error = OT_ERROR_NONE;
+    Message *message;
+
+    VerifyOrExit((message = FindFragmentPriorityMessage(aFragmentHeader)) != NULL, error = OT_ERROR_NOT_FOUND);
+    aPriority = message->GetPriority();
+
+exit:
+    return error;
+}
+
+void MeshForwarder::UpdateFragmentPriorityMessage(Lowpan::FragmentHeader &aFragmentHeader,
+                                                  uint8_t                 aFragmentLength,
+                                                  uint8_t                 aPriority)
+{
+    Message *message;
+
+    if (aFragmentHeader.GetDatagramOffset() == 0)
+    {
+        VerifyOrExit((message = GetInstance().GetMessagePool().New(Message::kTypeIp6, 0, aPriority)) != NULL);
+
+        message->SetDatagramTag(aFragmentHeader.GetDatagramTag());
+        message->SetTimeout(kReassemblyTimeout);
+        mFragmentPriorityQueue.Enqueue(*message);
+
+        if (!mUpdateTimer.IsRunning())
```

Comment:
I guess we need to just save the tag, priority and a timeout associated with this fragment. 

Allocating a `message` from pool will use at least one block of 128 bytes.  I remember in some earlier implementation you used a struct to store only these fields in a separate array.  

What do you think about the earlier model?

Reply:
You are right. From the perspective of consuming memory, the earlier model is better. I will change the implementation to the earlier model.